### DISCUSSION TITLE
[✨] Custom Class serialization

### DISCUSSION THREAD

**Discussion by swwind** [QwikDev/qwik-evolution#163] [Proposals For Qwik]
### Is your feature request related to a problem?

I'm using MongoDB and every request contains one or more `ObjectId` class object which can not be serialized.

And I'm lazy to write a bunch of `.map`s and `.toString()`s to convert those fields to string manually.

### Describe the solution you'd like

Write a bunch of `.map`s and `.toString()`s after every request is too terrible for me.

I'm looking for those solutions

1. Add a serializer to handle this `ObjectId` in `mongodb` (or in `bson`) just like `URL` and `Date`. (ideal but a waste for non-mongodb user)
2. Implement above serializer as a plugin for runtime. (possible?)
3. Support custom serialization for any class, such as using `toString()` or `qSerialize()` function automatically. (however, using `toString()` converts any class object into `string`, which requires additional type transform in `useLoader()` and `useAction`, etc.)

### Describe alternatives you've considered

n/a

**Comment by mikuhl-dev** [comment]
https://github.com/bluesky-social/atproto has a class object BlobRef that is also not serializable, would love to write a custom serializer instead of bugging the authors to change it.

**Comment by guzart** [comment]
Facing a similar challenge with Prisma returning a [Decimal object](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields#working-with-decimal). 

Would it be possible to support adding custom serializers via a vite plugin?

**Comment by mhevery** [comment]
Custom serializers is a complex problem, which is why we have been punting it. Perhaps you can send a PR to add the Decimal Object?

**Comment by wmertens** [comment]
I think most custom serializers can be avoided by putting methods in the store that on invocation check if the desired object exists and if not, create it with noSerialize before doing their own work.

The technique I describe [here](https://github.com/BuilderIO/qwik/issues/5087#issuecomment-1707185010) can be useful for that.

I don't think Qwik should ship with support for random libraries but maybe the $sync proposal can help with doing actual custom serialization

**Comment by gioboa** [comment]
We moved this issue to `qwik-evolution` repo to create a RFC discussion for this.
Here is our [Qwik RFC process](https://github.com/QwikDev/qwik-evolution?tab=readme-ov-file#qwik-rfc-process) thanks.

**Comment by wmertens** [comment]
closing this as we have `useSerializer$` in v2
